### PR TITLE
User Get

### PR DESCRIPTION
**Pull request by willsb**
Adds the `/me` endpoint to the UserClient.

Integration tests are not included because we need /me on v9.

I think we should merge as is and add integration tests once backend creates the endpoint on v9.

### CODE REVIEW

File: Toggl.Ultrawave/Network/HttpHeader.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+        public HeaderType Type { get; } 
+
+        public string Name { get; }
+
+        public string Value { get; }
+
+        public HttpHeader(string name, string value, HeaderType type = HeaderType.Other)
+        {
+			Type = type;
```
Comment (amulware) [inline Toggl.Ultrawave/Network/HttpHeader.cs]: Formatting.

File: Toggl.Ultrawave/TogglClient.cs [hunk 2]
```diff
@@ -1,18 +1,25 @@
...
+using Toggl.Ultrawave.Clients;
 using Toggl.Ultrawave.Network;
 
 namespace Toggl.Ultrawave
 {
     public sealed class TogglClient : ITogglClient
     {
         private readonly Endpoints endpoints;
+        private readonly IApiClient apiClient;
+        private readonly IJsonSerializer serializer;
```
Comment (amulware) [inline Toggl.Ultrawave/TogglClient.cs]: Why keep these around?

**Comment by amulware** [comment]
For reference, v9 now has `/me`. Please submit integration tests as separate PR.

File: Toggl.Ultrawave.Tests/Clients/UserClientTests.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            {
+                const string username = "susancalvin@psychohistorian.museum";
+                const string password = "theirobotmoviesucked123";
+                const string expectedHeader = "c3VzYW5jYWx2aW5AcHN5Y2hvaGlzdG9yaWFuLm11c2V1bTp0aGVpcm9ib3Rtb3ZpZXN1Y2tlZDEyMw==";
+
+                var userEndpoints = new Endpoints(ApiEnvironment.Staging).User;
+                var client = new UserClient(userEndpoints, apiClient, serializer);
+
+                await client.Get(username, password).Execute();
+                await apiClient.Received().Send(Arg.Is<Request>(request => checkRequest(request, expectedHeader)));
```
Comment (willsb) [inline Toggl.Ultrawave.Tests/Clients/UserClientTests.cs]: This is the API that NSubstitute gives us to check if method was called. The alternative would be creating extension methods, though we'd have to create one for each `Received()` call, since each one is pretty much unique...

Comment (willsb) [inline Toggl.Ultrawave.Tests/Clients/UserClientTests.cs]: I say we stick to calling `Received`, makes a heck lot more sense

File: Toggl.Ultrawave.Tests/Clients/UserClientTests.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+using NSubstitute;
+using Toggl.Ultrawave.Clients;
+using Toggl.Ultrawave.Network;
+using Xunit;
+
+namespace Toggl.Ultrawave.Tests
+{
+    public class UserClientTests
+    {
+        public class TheGetMethod
```
Comment (amulware) [inline Toggl.Ultrawave.Tests/Clients/UserClientTests.cs]: Should we also test that the request URL is correct? And what about the body of the request?

File: Toggl.Ultrawave.Tests/Clients/UserClientTests.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                var userEndpoints = new Endpoints(ApiEnvironment.Staging).User;
+                var client = new UserClient(userEndpoints, apiClient, serializer);
+
+                await client.Get(username, password).Execute();
+                await apiClient.Received().Send(Arg.Is<Request>(request => verifyAuthHeader(request, expectedHeader)));
+            }
+
+            private static bool verifyAuthHeader(Request request, string expectedHeader)
+            {
+                var authHeader = request.Headers.ToList().Single();
```
Comment (amulware) [inline Toggl.Ultrawave.Tests/Clients/UserClientTests.cs]: What happens here when we add more headers in the future? (which we will)

File: Toggl.Ultrawave/Clients/BaseClient.cs [hunk 6]
```diff
@@ -19,5 +19,24 @@ protected BaseClient(IApiClient apiClient, IJsonSerializer serializer)
             this.apiClient = apiClient;
             this.serializer = serializer;
         }
+
+        protected HttpHeader GetAuthHeader(string username, string password)
```
Comment (amulware) [inline Toggl.Ultrawave/Clients/BaseClient.cs]: Did we want to use strongly typed wrappers for username and password?

Comment (amulware) [inline Toggl.Ultrawave/Clients/BaseClient.cs]: (if so, feel free to create an issue, and we'll do it in a separate PR)

File: Toggl.Ultrawave/Network/HttpHeader.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
+namespace Toggl.Ultrawave.Network
+{
+    internal struct HttpHeader
+    {
+        internal enum HeaderType
```
Comment (amulware) [inline Toggl.Ultrawave/Network/HttpHeader.cs]: I think we had a nested type before and decided they should be public, right?

File: Toggl.Ultrawave.Tests/Clients/UserClientTests.cs (hunk 5 above)
Comment (willsb) [inline Toggl.Ultrawave.Tests/Clients/UserClientTests.cs]: Red. Green. Refactor.

File: Toggl.Ultrawave/Clients/BaseClient.cs (hunk 6 above)
Comment (willsb) [inline Toggl.Ultrawave/Clients/BaseClient.cs]: We do, yes. Will do so in a different PR

File: Toggl.Ultrawave/Network/HttpHeader.cs (hunk 7 above)
Comment (willsb) [inline Toggl.Ultrawave/Network/HttpHeader.cs]: Hasn't happened for types, but agreed

File: Toggl.Ultrawave.Tests/Clients/UserClientTests.cs (hunk 4 above)
Comment (willsb) [inline Toggl.Ultrawave.Tests/Clients/UserClientTests.cs]: IMO that's overkill already, since integration tests are the ones who'll really tell if the url/body is proper or not

Comment (amulware) [inline Toggl.Ultrawave.Tests/Clients/UserClientTests.cs]: Ok.

File: Toggl.Ultrawave.Tests/Clients/UserClientTests.cs (hunk 5 above)
Comment (amulware) [inline Toggl.Ultrawave.Tests/Clients/UserClientTests.cs]: Ok.
